vm/env.go

```diff
@@ -160,15 +160,18 @@ func (e *Env) Get(k string) (reflect.Value, error) {
 // Set modifies value which specified as symbol. It goes to upper scope until
 // found or returns error.
 func (e *Env) Set(k string, v interface{}) error {
-	e.Lock()
-	defer e.Unlock()
-
-	if _, ok := e.env[k]; ok {
-		val, ok := v.(reflect.Value)
+	e.RLock()
+	_, ok := e.env[k]
+	e.RUnlock()
+	if ok {
+		var val reflect.Value
+		val, ok = v.(reflect.Value)
 		if !ok {
 			val = reflect.ValueOf(v)
 		}
+		e.Lock()
 		e.env[k] = val
+		e.Unlock()
 		return nil
```

Comment:
I think this possible be race condition.

Reply:
I think situation can arise when two goroutines try to use Set and place between using RUnlock() and e.Lock() could be accessed by two subroutines, right?

Reply:
Yes.

Reply:
The read lock is a check if the variable is in the same env. If one did a define and set at the same time, one might set the variable in a different env than expected. This could happen even if there was a write lock on the whole function. If there are some Go routines running defines and some running sets on the same variable name in different parent/child env, there will always be a race condition on  which env or more than one env that variable name gets defined/set.

Example With pEnv & cEnv (even if set had global write lock):
GoR1: pEnv.Define A
GoR1: cEnv.Define A
GoR2: cEnv.Set A
Result: A set in cEnv 

GoR1: pEnv.Define A
GoR2: cEnv.Set A
GoR1: cEnv.Define A
Result: A set in pEnv

There should be no deadlock from this code change

Reply:
In this part, `ok` possibly become true at multiple groutine in same time?

https://github.com/mattn/anko/pull/81/files/574a908f79c43f50ce76f71ad3ede3dae2ff7cd4#diff-12238057aa85e299611b1bd13ffa555fR166

Reply:
Yes, if the variable is not Define yet, it would keep being false.

The difference between this updated Set function and the previous one is that in the previous version once Set gets its write lock any Define calls don't affect it. In this version, the Define call still affects the out come of the Set call in progress, depending on where the code is in the Set function. Both the previous version and this version have race conditions, depending on which function gets the locks first.  The winner of that race will affect if the variable will be set in this env scope or the parent scope, if there is a parent env.

So lets say two scripts are running in the same env with no parent scope, both are trying to set the same variable at the same time. Both get past the Set read lock and show it is false, and then they both return the Unknown symbol error. So both scripts call Define at the same time. Both hit the Define lock at the same time but only one of them will win. After one of the scripts sets the variable, the other one will override the variable. Both had to make the extra call to Define, but that should not be the usual case.

Reply:
```
#!anko

func foo() {
  var f = 1
  func() {
    #var f = 0
    f = 2
  }()
  print(f)
}

foo();
```
When this code is run, it show 2. And when un-comment the line, it show 1. Set will follow the parent scope to find f if exists in the scope. If the variable f is asynchronously set with Define on another goroutine, the Set will reference the parent 's env.

Reply:
Are you you saying something is not working correctly with that code? If so, could you please point it out to me?

Reply:
Added TestRaceDefineAndSetSameVariable, which shows the define/set race.